ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -240,16 +241,21 @@ private void pollQuotaExceedAllowedRequestsIfAny(List<RequestInfo> requestsToSen
         requestQueue.get(quotaResource).removeFirst();
         if (requestQueue.get(quotaResource).isEmpty()) {
           requestQueue.remove(quotaResource);
+          // This quotaResource has exhausted, we should remove it from next for loop
+          quotaResourcesRemoved.add(quotaResource);
         }
       }
+      quotaResources.removeAll(quotaResourcesRemoved);
```

Comment:
if the quotaResource is exhausted, we need to remove it from quotaResources list so next time we do the for loop in line 226, we don't have to deal with it anymore.